File: python/visualize.py
import colorsys
import numpy as np
import pygame
import torch
from snake_env import SnakeEnv
from network import forward_with_activations

# ── Layout ──────────────────────────────────────────────────────────────────
CELL   = 28
GRID   = 20
GAME_W = GRID * CELL   # 560
GAME_H = GRID * CELL   # 560
HUD_H  = 50
GAP    = 10

PANEL_X = GAME_W + GAP  # 570
PANEL_W = 370
WIN_W   = PANEL_X + PANEL_W  # 940
WIN_H   = GAME_H + HUD_H     # 610

# NN node x positions (absolute)
INP_X  = PANEL_X + 50
HID_X  = PANEL_X + 185
OUT_X  = PANEL_X + 310
LOUT_X = PANEL_X + 327

# Node radii
R_INP = 5
R_HID = 3   # smaller — 64 nodes need tighter packing
R_OUT = 9

# Vertical node positions
def _ys(n, top=18, bottom=GAME_H - 18):
    if n == 1:
        return [(top + bottom) // 2]
    return [int(top + i * (bottom - top) / (n - 1)) for i in range(n)]

INP_YS = _ys(20)
HID_YS = _ys(64)
OUT_YS = _ys(4, top=115, bottom=495)

SENSOR_DIRS = ["N", "NE", "E", "SE", "S", "SW", "W", "NW"]
OUT_LABELS  = ["↑", "→", "↓", "←"]

GROUPS = [
    ("WALL", (35, 15, 15), (230,  75,  75)),
    ("BODY", (35, 28, 10), (230, 170,  40)),
]
FOOD_DARK   = (10, 20, 40)
FOOD_BRIGHT = (60, 160, 255)
FOOD_LABELS = ["↑", "↓", "←", "→"]

HID_DARK   = (22, 22, 22)
HID_BRIGHT = (255, 200, 50)

BG       = (15,  15,  15)
GRID_C   = (25,  25,  25)
TEXT_C   = (200, 200, 200)
PANEL_BG = (20,  20,  20)
SEP_C    = (50,  50,  50)
DIM_C    = (70,  70,  80)
ACT_POS  = ( 80, 140, 255)
ACT_NEG  = (255,  80,  80)


def _lerp(a, b, t):
    t = max(0.0, min(1.0, float(t)))
    return (int(a[0] + (b[0]-a[0])*t), int(a[1] + (b[1]-a[1])*t), int(a[2] + (b[2]-a[2])*t))


def _rank_color(rank: int, n: int = 10) -> tuple:
    """Bright saturated green (rank 0) → dim desaturated (rank n-1)."""
    t = rank / max(n - 1, 1)
    h = 0.38  # green-cyan hue
    s = max(0.12, 1.0 - t * 0.85)
    v = max(0.28, 1.0 - t * 0.68)
    r, g, b = colorsys.hsv_to_rgb(h, s, v)
    return (int(r * 255), int(g * 255), int(b * 255))


class Visualizer:
    def __init__(self, grid_size: int = GRID, cell: int = CELL):
        self.grid_size = grid_size
        self.cell = cell

        pygame.init()
        self.screen = pygame.display.set_mode((WIN_W, WIN_H))
        pygame.display.set_caption("Snake AI")
        self.font_md = pygame.font.SysFont("monospace", 15)
        self.font_sm = pygame.font.SysFont("monospace", 11)
        self.clock   = pygame.time.Clock()

        self.conn_surf = pygame.Surface((WIN_W, WIN_H), pygame.SRCALPHA)

    def play_top_n(
        self,
        params_list: list,
        device: torch.device,
        generation: int,
        fps: int = 15,
    ) -> int:
        """
        Play n snake episodes simultaneously (ordered best→worst).
        Snakes are overlaid on one grid; best = most saturated color.
        NN panel shows the best alive snake. Returns best score, -1 on quit.
        """
        n = len(params_list)
        envs    = [SnakeEnv(self.grid_size) for _ in range(n)]
        obs_arr = [env.reset() for env in envs]
        params  = [p.to(device) for p in params_list]
        alive   = [True] * n
        colors  = [_rank_color(i, n) for i in range(n)]
        best_acts = None

        while True:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    return -1
                if event.type == pygame.KEYDOWN and event.key == pygame.K_q:
                    return -1

            step_acts = []
            for i in range(n):
                if not alive[i]:
                    step_acts.append(None)
                    continue
                obs_t = torch.tensor(obs_arr[i], device=device)
                with torch.no_grad():
                    acts = forward_with_activations(params[i], obs_t)
                step_acts.append(acts)
                obs_arr[i], _, done = envs[i].step(acts["action"])
                if done:
                    alive[i] = False

            # NN display tracks the highest-ranked snake still alive
            for i in range(n):
                if alive[i]:
                    best_acts = step_acts[i]
                    break

            self.screen.fill(BG)
            self._draw_game_multi(envs, alive, colors)
            if best_acts is not None:
                self._draw_panel(best_acts)
            self._draw_hud_multi(envs, alive, colors, generation)
            pygame.display.flip()
            self.clock.tick(fps)

            if not any(alive):
                pygame.time.wait(700)
                return max(env.score for env in envs)

    # ── Game rendering ───────────────────────────────────────────────────────

    def _draw_game_multi(self, envs, alive, colors):
        c = self.cell

        for i in range(self.grid_size + 1):
            pygame.draw.line(self.screen, GRID_C, (i*c, 0), (i*c, GAME_H))
            pygame.draw.line(self.screen, GRID_C, (0, i*c), (GAME_W, i*c))

        # Food dots — small, tinted to match each snake
        for i, env in enumerate(envs):
            if not alive[i]:
                continue
            fr, fc = env.food
            r, g, b = colors[i]
            pygame.draw.circle(
                self.screen, (r//2, g//2, b//2),
                (fc*c + c//2, fr*c + c//2), 4,
            )

        # Snakes — draw worst first so best renders on top
        for i in range(len(envs) - 1, -1, -1):
            if not alive[i]:
                continue
            clr = colors[i]
            head = tuple(min(255, int(x * 1.45)) for x in clr)
            for idx, (r, col) in enumerate(envs[i].snake):
                pygame.draw.rect(
                    self.screen, head if idx == 0 else clr,
                    (col*c + 1, r*c + 1, c - 2, c - 2),
                )

    # ── NN panel ─────────────────────────────────────────────────────────────

    def _draw_panel(self, acts: dict):
        pygame.draw.rect(self.screen, PANEL_BG, (PANEL_X, 0, PANEL_W, GAME_H + HUD_H))
        pygame.draw.line(self.screen, SEP_C, (PANEL_X - 1, 0), (PANEL_X - 1, WIN_H))

        self.screen.blit(
            self.font_sm.render("NEURAL  NETWORK", True, DIM_C),
            (PANEL_X + 110, 4),
        )
        for text, x in [("INPUT", INP_X-18), ("HIDDEN", HID_X-20), ("OUTPUT", OUT_X-22)]:
            self.screen.blit(self.font_sm.render(text, True, DIM_C), (x, GAME_H - 14))

        self._draw_connections(acts)
        self._draw_neurons(acts)

    def _draw_connections(self, acts: dict):
        self.conn_surf.fill((0, 0, 0, 0))

        inp = acts["inputs"]
        hid = acts["hidden"]
        w1  = acts["w1"]
        w2  = acts["w2"]

        for i in range(20):
            for j in range(16):
                contrib = abs(float(w1[i, j]) * float(inp[i]))
                alpha = min(210, int(contrib * 700))
                if alpha < 8:
                    continue
                clr = (*ACT_POS, alpha) if w1[i, j] > 0 else (*ACT_NEG, alpha)
                pygame.draw.line(self.conn_surf, clr, (INP_X, INP_YS[i]), (HID_X, HID_YS[j]))

        for j in range(16):
            for k in range(4):
                contrib = abs(float(w2[j, k]) * float(hid[j]))
                alpha = min(210, int(contrib * 700))
                if alpha < 8:
                    continue
                clr = (*ACT_POS, alpha) if w2[j, k] > 0 else (*ACT_NEG, alpha)
                pygame.draw.line(self.conn_surf, clr, (HID_X, HID_YS[j]), (OUT_X, OUT_YS[k]))

        self.screen.blit(self.conn_surf, (0, 0))

    def _draw_neurons(self, acts: dict):
        inp    = acts["inputs"]
        hid    = acts["hidden"]
        out    = acts["output"]
        action = acts["action"]

        hid_max = max(float(hid.max()), 1e-6)
        exp_o   = np.exp(out - out.max())
        softmax = exp_o / exp_o.sum()

        for i in range(16):
            g = i // 8
            _, dark, bright = GROUPS[g]
            pygame.draw.circle(self.screen, _lerp(dark, bright, float(inp[i])),
                               (INP_X, INP_YS[i]), R_INP)
            self.screen.blit(
                self.font_sm.render(SENSOR_DIRS[i % 8], True, (65, 65, 75)),
                (INP_X - 24, INP_YS[i] - 5),
            )

        # Food direction nodes 16-19 (all non-negative)
        for j in range(4):
            fi = 16 + j
            pygame.draw.circle(self.screen, _lerp(FOOD_DARK, FOOD_BRIGHT, float(inp[fi])),
                               (INP_X, INP_YS[fi]), R_INP)
            self.screen.blit(
                self.font_sm.render(FOOD_LABELS[j], True, (65, 65, 75)),
                (INP_X - 14, INP_YS[fi] - 5),
            )

        for g, (name, _, bright) in enumerate(GROUPS):
            mid_y = (INP_YS[g*8] + INP_YS[g*8+7]) // 2
            self.screen.blit(self.font_sm.render(name, True, bright), (PANEL_X+3, mid_y-5))
        food_mid = (INP_YS[16] + INP_YS[19]) // 2
        self.screen.blit(self.font_sm.render("FOOD", True, FOOD_BRIGHT), (PANEL_X+3, food_mid-5))

        for j in range(64):
            t = float(hid[j]) / hid_max
            pygame.draw.circle(self.screen, _lerp(HID_DARK, HID_BRIGHT, t),
                               (HID_X, HID_YS[j]), R_HID)

        for k in range(4):
            clr = (30, 230, 90) if k == action else (int(softmax[k]*160),)*3
            pygame.draw.circle(self.screen, clr, (OUT_X, OUT_YS[k]), R_OUT)
            self.screen.blit(self.font_md.render(OUT_LABELS[k], True, TEXT_C),
                             (LOUT_X, OUT_YS[k]-8))

    # ── HUD ──────────────────────────────────────────────────────────────────

    def _draw_hud_multi(self, envs, alive, colors, generation):
        best_score = max(env.score for env in envs)
        n_alive    = sum(alive)
        n          = len(envs)

        hud = self.font_md.render(
            f"gen {generation}   best score {best_score}   alive {n_alive}/{n}",
            True, TEXT_C,
        )
        self.screen.blit(hud, (8, GAME_H + 16))

        # Color legend: up to 40 swatches before the label would overflow
        lx = 8
        max_swatches = min(len(colors), 40)
        for clr in colors[:max_swatches]:
            pygame.draw.rect(self.screen, clr, (lx, GAME_H + 36, 10, 6))
            lx += 12
        label = "best→worst" if len(colors) <= 40 else f"best→worst ({len(colors)} snakes)"
        self.screen.blit(self.font_sm.render(label, True, DIM_C), (lx + 4, GAME_H + 33))

        # Connection legend
        lx2 = PANEL_X + 8
        ly  = GAME_H + 12
        for clr, txt in [(ACT_POS, "+weight"), (ACT_NEG, "-weight")]:
            pygame.draw.line(self.screen, clr, (lx2, ly+5), (lx2+18, ly+5), 2)
            self.screen.blit(self.font_sm.render(txt, True, DIM_C), (lx2+22, ly))
            lx2 += 90

    def close(self):
        pygame.quit()

File: rust/src/visualize.rs
use macroquad::prelude::*;

use crate::network::{forward_with_activations, IN_SIZE, H_SIZE, OUT_SIZE};
use crate::snake_env::SnakeEnv;

// ── Layout constants ─────────────────────────────────────────────────────────
pub const CELL:    f32 = 28.0;
pub const GRID:    i32 = 20;
pub const GAME_W:  f32 = CELL * GRID as f32;   // 560
pub const GAME_H:  f32 = GAME_W;               // 560
pub const HUD_H:   f32 = 50.0;
pub const PANEL_X: f32 = GAME_W + 10.0;        // 570
pub const WIN_W:   f32 = PANEL_X + 370.0;      // 940
pub const WIN_H:   f32 = GAME_H + HUD_H;       // 610

const INP_X:  f32 = PANEL_X + 50.0;
const HID_X:  f32 = PANEL_X + 185.0;
const OUT_X:  f32 = PANEL_X + 310.0;
const LOUT_X: f32 = PANEL_X + 323.0;

const R_INP: f32 = 5.0;
const R_HID: f32 = 3.0;  // smaller — 64 nodes need tighter packing
const R_OUT: f32 = 9.0;

const SENSOR_DIRS: [&str; 8] = ["N","NE","E","SE","S","SW","W","NW"];
const OUT_LABELS:  [&str; 4] = ["^", ">", "v", "<"];

// group (name, dark rgb, bright rgb)
const WALL_DARK:  (f32,f32,f32) = (0.14, 0.06, 0.06);
const WALL_BRIGHT:(f32,f32,f32) = (0.90, 0.29, 0.29);
const BODY_DARK:  (f32,f32,f32) = (0.14, 0.11, 0.04);
const BODY_BRIGHT:(f32,f32,f32) = (0.90, 0.67, 0.16);
const FOOD_DARK:  (f32,f32,f32) = (0.04, 0.08, 0.16);
const FOOD_BRIGHT:(f32,f32,f32) = (0.24, 0.63, 1.00);
const HID_DARK:   (f32,f32,f32) = (0.09, 0.09, 0.09);
const HID_BRIGHT: (f32,f32,f32) = (1.00, 0.78, 0.20);

fn lerp_col(a: (f32,f32,f32), b: (f32,f32,f32), t: f32) -> Color {
    let t = t.clamp(0.0, 1.0);
    Color::new(a.0+(b.0-a.0)*t, a.1+(b.1-a.1)*t, a.2+(b.2-a.2)*t, 1.0)
}

fn rank_color(rank: usize, n: usize) -> Color {
    let t = rank as f32 / (n.saturating_sub(1).max(1)) as f32;
    let h = 0.38f32;
    let s = (1.0 - t * 0.85).max(0.12);
    let v = (1.0 - t * 0.68).max(0.28);
    hsv(h, s, v)
}

// standard HSV → RGBA (macroquad helper)
fn hsv(h: f32, s: f32, v: f32) -> Color {
    let i = (h * 6.0).floor() as i32;
    let f = h * 6.0 - i as f32;
    let p = v * (1.0 - s);
    let q = v * (1.0 - f * s);
    let t = v * (1.0 - (1.0 - f) * s);
    let (r, g, b) = match i % 6 {
        0 => (v, t, p),
        1 => (q, v, p),
        2 => (p, v, t),
        3 => (p, q, v),
        4 => (t, p, v),
        _ => (v, p, q),
    };
    Color::new(r, g, b, 1.0)
}

fn make_ys(n: usize, top: f32, bottom: f32) -> Vec<f32> {
    if n == 1 { return vec![(top + bottom) / 2.0]; }
    (0..n).map(|i| top + i as f32 * (bottom - top) / (n - 1) as f32).collect()
}

pub struct Visualizer {
    pub grid_size: i32,
    inp_ys: Vec<f32>,
    hid_ys: Vec<f32>,
    out_ys: Vec<f32>,
}

impl Visualizer {
    pub fn new(grid_size: i32) -> Self {
        Self {
            grid_size,
            inp_ys: make_ys(IN_SIZE, 18.0, 542.0),
            hid_ys: make_ys(H_SIZE,  18.0, 542.0),
            out_ys: make_ys(OUT_SIZE, 115.0, 495.0),
        }
    }

    /// Play n snakes simultaneously. Returns best score, or -1 if user quits.
    pub async fn play_top_n(
        &self,
        params_list: &[Vec<f32>],
        generation:  usize,
        fps:         f64,
    ) -> i32 {
        let n       = params_list.len();
        let mut envs: Vec<SnakeEnv> = (0..n).map(|_| SnakeEnv::new(self.grid_size, vec![])).collect();
        let mut obs:  Vec<Vec<f32>> = envs.iter_mut().map(|e| e.reset()).collect();
        let mut alive = vec![true; n];
        let colors: Vec<Color> = (0..n).map(|i| rank_color(i, n)).collect();
        let mut best_acts = None;

        let step_dt = 1.0 / fps;
        let mut last_step = get_time();

        loop {
            // ── Input ──
            if is_key_pressed(KeyCode::Q) || is_key_pressed(KeyCode::Escape) {
                return -1;
            }

            // ── Step snakes at target fps ──
            let now = get_time();
            if now - last_step >= step_dt {
                last_step = now;
                for i in 0..n {
                    if !alive[i] { continue; }
                    let acts = forward_with_activations(&params_list[i], &obs[i]);
                    let (new_obs, done) = envs[i].step(acts.action);
                    obs[i] = new_obs;
                    if i == 0 || best_acts.is_none() {
                        // track best-alive for NN display (first alive = rank 0)
                        if alive[i] { best_acts = Some(acts); }
                    }
                    if done { alive[i] = false; }
                }
                // find best still alive for NN display
                best_acts = None;
                for i in 0..n {
                    if alive[i] {
                        best_acts = Some(forward_with_activations(&params_list[i], &obs[i]));
                        break;
                    }
                }
            }

            // ── Draw ──
            clear_background(Color::from_rgba(15, 15, 15, 255));
            self.draw_game(&envs, &alive, &colors);
            if let Some(ref acts) = best_acts {
                self.draw_panel(acts);
            }
            self.draw_hud(&envs, &alive, &colors, generation);

            next_frame().await;

            if alive.iter().all(|&a| !a) {
                // brief pause before returning
                let end = get_time() + 0.7;
                while get_time() < end { next_frame().await; }
                let best = envs.iter().map(|e| e.score).max().unwrap_or(0);
                return best as i32;
            }
        }
    }

    // ── Game panel ───────────────────────────────────────────────────────────

    fn draw_game(&self, envs: &[SnakeEnv], alive: &[bool], colors: &[Color]) {
        let c = CELL;

        // Grid
        let grid_col = Color::from_rgba(25, 25, 25, 255);
        for i in 0..=(self.grid_size as usize) {
            let x = i as f32 * c;
            let y = i as f32 * c;
            draw_line(x, 0.0, x, GAME_H, 1.0, grid_col);
            draw_line(0.0, y, GAME_W, y, 1.0, grid_col);
        }

        // Food — draw worst-first so best food is on top
        for i in (0..envs.len()).rev() {
            if !alive[i] { continue; }
            let (fr, fc) = envs[i].food;
            let col = colors[i];
            let food_col = Color::new(col.r * 0.5, col.g * 0.5, col.b * 0.5, 1.0);
            draw_circle(fc as f32 * c + c / 2.0, fr as f32 * c + c / 2.0, 4.0, food_col);
        }

        // Snakes — worst first so best renders on top
        for i in (0..envs.len()).rev() {
            if !alive[i] { continue; }
            let col = colors[i];
            let head = Color::new(
                (col.r * 1.45).min(1.0),
                (col.g * 1.45).min(1.0),
                (col.b * 1.45).min(1.0),
                1.0,
            );
            for (idx, &(r, col_pos)) in envs[i].snake.iter().enumerate() {
                let clr = if idx == 0 { head } else { col };
                draw_rectangle(col_pos as f32 * c + 1.0, r as f32 * c + 1.0, c - 2.0, c - 2.0, clr);
            }
        }
    }

    // ── NN panel ─────────────────────────────────────────────────────────────

    fn draw_panel(&self, acts: &crate::network::ForwardResult) {
        // Panel background
        draw_rectangle(PANEL_X, 0.0, WIN_W - PANEL_X, WIN_H, Color::from_rgba(20, 20, 20, 255));
        draw_line(PANEL_X - 1.0, 0.0, PANEL_X - 1.0, WIN_H, 1.0, Color::from_rgba(50, 50, 50, 255));

        draw_text("NEURAL NETWORK", PANEL_X + 100.0, 14.0, 14.0, Color::from_rgba(70, 70, 80, 255));
        let dim = Color::from_rgba(70, 70, 80, 255);
        draw_text("INPUT",  INP_X - 18.0, GAME_H - 4.0, 12.0, dim);
        draw_text("HIDDEN", HID_X - 22.0, GAME_H - 4.0, 12.0, dim);
        draw_text("OUTPUT", OUT_X - 24.0, GAME_H - 4.0, 12.0, dim);

        self.draw_connections(acts);
        self.draw_neurons(acts);
    }

    fn draw_connections(&self, acts: &crate::network::ForwardResult) {
        let inp = &acts.inputs;
        let hid = &acts.hidden;
        let w1  = &acts.w1;
        let w2  = &acts.w2;

        for i in 0..IN_SIZE {
            for j in 0..H_SIZE {
                let contrib = (w1[i][j] * inp[i]).abs();
                let alpha = (contrib * 0.7).min(0.82);
                if alpha < 0.03 { continue; }
                let (r, g, b) = if w1[i][j] > 0.0 { (0.31, 0.55, 1.0) } else { (1.0, 0.31, 0.31) };
                draw_line(INP_X, self.inp_ys[i], HID_X, self.hid_ys[j], 1.0,
                    Color::new(r, g, b, alpha));
            }
        }

        for j in 0..H_SIZE {
            for k in 0..OUT_SIZE {
                let contrib = (w2[j][k] * hid[j]).abs();
                let alpha = (contrib * 0.7).min(0.82);
                if alpha < 0.03 { continue; }
                let (r, g, b) = if w2[j][k] > 0.0 { (0.31, 0.55, 1.0) } else { (1.0, 0.31, 0.31) };
                draw_line(HID_X, self.hid_ys[j], OUT_X, self.out_ys[k], 1.0,
                    Color::new(r, g, b, alpha));
            }
        }
    }

    fn draw_neurons(&self, acts: &crate::network::ForwardResult) {
        let inp    = &acts.inputs;
        let hid    = &acts.hidden;
        let out    = &acts.output;
        let action = acts.action;

        let hid_max = hid.iter().cloned().fold(1e-6f32, f32::max);

        // Softmax for output display
        let out_max = out.iter().cloned().fold(f32::NEG_INFINITY, f32::max);
        let exp: Vec<f32> = out.iter().map(|&v| (v - out_max).exp()).collect();
        let exp_sum: f32  = exp.iter().sum();
        let softmax: Vec<f32> = exp.iter().map(|&e| e / exp_sum).collect();

        let dim_text = Color::from_rgba(65, 65, 75, 255);

        // Input neurons
        for i in 0..IN_SIZE {
            let clr = if i < 8 {
                lerp_col(WALL_DARK, WALL_BRIGHT, inp[i])
            } else if i < 16 {
                lerp_col(BODY_DARK, BODY_BRIGHT, inp[i])
            } else {
                lerp_col(FOOD_DARK, FOOD_BRIGHT, inp[i])  // already non-negative
            };
            draw_circle(INP_X, self.inp_ys[i], R_INP, clr);

            let food_lbls = ["^", "v", "<", ">"];
            let lbl = if i < 16 { SENSOR_DIRS[i % 8] } else { food_lbls[i - 16] };
            draw_text(lbl, INP_X - 14.0, self.inp_ys[i] + 4.0, 11.0, dim_text);
        }

        // Group labels
        let wall_mid = (self.inp_ys[0] + self.inp_ys[7]) / 2.0;
        let body_mid = (self.inp_ys[8] + self.inp_ys[15]) / 2.0;
        let food_mid = (self.inp_ys[16] + self.inp_ys[19]) / 2.0;
        draw_text("WALL", PANEL_X + 2.0, wall_mid + 4.0, 11.0, Color::from_rgba(230, 75, 75, 255));
        draw_text("BODY", PANEL_X + 2.0, body_mid + 4.0, 11.0, Color::from_rgba(230, 170, 40, 255));
        draw_text("FOOD", PANEL_X + 2.0, food_mid + 4.0, 11.0, Color::from_rgba(60, 160, 255, 255));

        // Hidden neurons — draw every other one to avoid overdraw at 64 nodes
        for j in 0..H_SIZE {
            let t = hid[j] / hid_max;
            draw_circle(HID_X, self.hid_ys[j], R_HID, lerp_col(HID_DARK, HID_BRIGHT, t));
        }
        let _h = H_SIZE; // suppress unused warning

        // Output neurons
        for k in 0..OUT_SIZE {
            let clr = if k == action {
                Color::from_rgba(30, 230, 90, 255)
            } else {
                let v = (softmax[k] * 160.0) as u8;
                Color::from_rgba(v, v, v, 255)
            };
            draw_circle(OUT_X, self.out_ys[k], R_OUT, clr);
            draw_text(OUT_LABELS[k], LOUT_X, self.out_ys[k] + 5.0, 16.0, WHITE);
        }
    }

    // ── HUD ──────────────────────────────────────────────────────────────────

    fn draw_hud(&self, envs: &[SnakeEnv], alive: &[bool], colors: &[Color], generation: usize) {
        let best_score = envs.iter().map(|e| e.score).max().unwrap_or(0);
        let n_alive    = alive.iter().filter(|&&a| a).count();
        let n          = envs.len();

        let txt = format!("gen {}   best score {}   alive {}/{}", generation, best_score, n_alive, n);
        draw_text(&txt, 8.0, GAME_H + 22.0, 16.0, Color::from_rgba(200, 200, 200, 255));

        // Color swatch legend
        let max_swatches = n.min(40);
        let mut lx = 8.0f32;
        for i in 0..max_swatches {
            draw_rectangle(lx, GAME_H + 36.0, 10.0, 6.0, colors[i]);
            lx += 12.0;
        }
        let lbl = if n <= 40 { "best->worst".to_string() } else { format!("best->worst ({n})") };
        draw_text(&lbl, lx + 4.0, GAME_H + 44.0, 11.0, Color::from_rgba(70, 70, 80, 255));
    }
}
